durn/actions.go

```diff
@@ -11,24 +10,86 @@ import (
 	"durn2.0/util"
 )
 
-// AddValidVoters checks that provided voters are entered as valid emails, and if that
-// is the case inserts them into the database
-func AddValidVoters(ctx context.Context, voters []models.Voter) error {
+// AddValidVoters checks that provided voters are entered as valid emails, and
+// if that is the case inserts them into the database.
+// Returns all voters for which it failed
+func AddValidVoters(ctx context.Context, voters []models.Voter) ([]models.Voter, error) {
+	dbConn := db.TakeDB()
+	defer db.ReleaseDB()
+
 	mailRegex := "[^@]+@kth\\.se"
+	votersToInsert := []db.ValidVoter{}
+	currentVoters := []db.ValidVoter{}
+	failedVoters := []models.Voter{}
+
+	voterExists := func(email models.Voter) bool {
+		for _, voter := range currentVoters {
+			if voter.Email == string(email) {
+				return true
+			}
+		}
+		return false
+	}
+
+	if result := dbConn.Find(&currentVoters); result.Error != nil {
+		rl.Warning(ctx, result.Error.Error())
+		return nil, util.ServerError("An internal server error occurred")
```

Comment:
My though was that the actual error is logged, and this is all that is returned to the api-user in order to avid leakage of information. I'm not 100% sure that this is a correct reasoning though, so I will look through this at other places as well.

Reply:
Since it is open source you probably don't leak that much info by saying where it failed. A user may be able to see what they did wrong directly if they know where it failed. Fair points though.